### PR TITLE
Workflow : Add nightly release of NumPy in linux workflows 

### PR DESCRIPTION
**Pull request by V0lantis**
#### Reference issue
This is a follow up to my PR #13876 which was failing once we merged it into master. I tested it on local and on my own fork and it seems to be working, at least I hope. You can found the workflow [here](https://github.com/V0lantis/scipy/runs/2512938768)

#### What does this implement/fix?
This PR takes the nightly release of Numpy to have the latest changes.

EDIT : 
@BvB93 can you look to the `# type: ignore[attr-defined]` statement for `TestCase` import (for example in `scipy/optimize/_trustregion_constr/tests/test_qp_subproblem.py`). I had the following errors : 

```
scipy/optimize/tests/test_differentiable_functions.py:3: error: Module 'numpy.testing' has no attribute 'TestCase'  [attr-defined]
scipy/optimize/tests/test_minimize_constrained.py:7: error: Module 'numpy.testing' has no attribute 'TestCase'  [attr-defined]
scipy/optimize/tests/test_hessian_update_strategy.py:4: error: Module 'numpy.testing' has no attribute 'TestCase'  [attr-defined]
scipy/optimize/tests/test_constraints.py:3: error: Module 'numpy.testing' has no attribute 'TestCase'  [attr-defined]
```

I wonder If it comes from numpy?

### CODE REVIEW

**Comment by BvB93** [comment]
Yup, good call. Turns out I missed objects in `np.testing`.
https://github.com/numpy/numpy/pull/18919 should fix the issue.

**Comment by rgommers** [comment]
Thanks @V0lantis, and thanks for the quick fix @BvB93.

The numpy nightlies are built once a week, on Friday night typically, so a build which includes the fix will show up at https://pypi.anaconda.org/scipy-wheels-nightly/simple/numpy/ on Saturday. Easiest way to check is to use the version hash (everything after `.g` in the wheel name, don't ask me why the `g` is there ....). @V0lantis can you update this PR on Saturday, removing the `# type: ignore`'s? And then I think this is good to merge.

**Comment by V0lantis** [comment]
I have made the changes and tested upon the commit from @BvB93. it works. I pushed and next Saturday I will merge master to trigger a new workflows.

**Comment by andyfaff** [comment]
Why is the mypy test being run before the actual test run itself? I'd like to see it the other way around so that we can see any test failures first, any typing failures are less consequential.

In addition, 3.10 has now been released to beta. This CI run was originally written to identify issues with as-yet unreleased Python. Therefore, I think we need to amend the "Nightly" Action item to run off Python 3.10. If this is going to cause issues with the mypy run, then either we need to create a new Action, or move the mypy run to another Action item.

**Comment by rgommers** [comment]
> Why is the mypy test being run before the actual test run itself? I'd like to see it the other way around so that we can see any test failures first, any typing failures are less consequential.

Probably because mypy is >10x faster to run. Both need to pass, so it doesn't matter much.

> In addition, 3.10 has now been released to beta. This CI run was originally written to identify issues with as-yet unreleased Python. Therefore, I think we need to amend the "Nightly" Action item to run off Python 3.10. If this is going to cause issues with the mypy run, then either we need to create a new Action, or move the mypy run to another Action item.

+1 for upgrading to 3.10. It's orthogonal to this PR, but if GitHub Actions has 3.10 available it's a one-liner change. Can you try to include it here @V0lantis? If it doesn't work easily, let's do it in a separate PR.

**Comment by V0lantis** [comment]
> +1 for upgrading to 3.10. It's orthogonal to this PR, but if GitHub Actions has 3.10 available it's a one-liner change. Can you try to include it here @V0lantis? If it doesn't work easily, let's do it in a separate PR.

Yeap, going to try this

**Comment by V0lantis** [comment]
> Why is the mypy test being run before the actual test run itself? I'd like to see it the other way around so that we can see any test failures first, any typing failures are less consequential.
> 
> In addition, 3.10 has now been released to beta. This CI run was originally written to identify issues with as-yet unreleased Python. Therefore, I think we need to amend the "Nightly" Action item to run off Python 3.10. If this is going to cause issues with the mypy run, then either we need to create a new Action, or move the mypy run to another Action item.

Unfortunately, we cannot because numpy still doesn't have a release for Python3.10 : 
``
ERROR: Could not find a version that satisfies the requirement numpy (from versions: none)
9
ERROR: No matching distribution found for numpy
10
Error: Process completed with exit code 1.
``` 

You can check the workflow [here ](https://github.com/V0lantis/scipy/runs/2526406067?check_suite_focus=true#step:6:10)

**Comment by rgommers** [comment]
Ah yes, let's open a separate issue for upgrading the nightly to py3.10, and then work on first getting NumPy nightly wheels for 3.10 up. Not difficult, but it's an extra CI job there I think.

**Comment by andyfaff** [comment]
The nightly CI used to build numpy from source, I don't think it took too long.

**Comment by rgommers** [comment]
Ah yes, going back to building from source should be okay as well. Indeed much faster than building scipy itself, so not a big issue.

**Comment by V0lantis** [comment]
Ok so should I do an inplace build of numpy in the Nightly python job ?

**Comment by rgommers** [comment]
No, a regular `pip install .` is preferred. Or, given you don't have a numpy clone yet:
```
pip install https://github.com/numpy/numpy/archive/refs/heads/main.zip
```

**Comment by V0lantis** [comment]
> pip install https://github.com/numpy/numpy/archive/refs/heads/main.zip

This doesn't work, so I made a workaround. It is not pretty nor fast but it works fine. I tried with 
`wget -r -np -R "index.html*"  --cut-dirs=6 https://github.com/numpy/numpy/archive/refs/heads/main.zip
` but it added a lot of overhead so I went back to this solution which works fine. 

The [failure](https://github.com/V0lantis/scipy/runs/2531089899?check_suite_focus=true) with wget laid in the version of numpy with wget was : 

```
_get_build_requires
2046
      self.run_setup()
2047
    File "/tmp/pip-build-env-15otxrx6/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 248, in run_setup
2048
      super(_BuildMetaLegacyBackend,
2049
    File "/tmp/pip-build-env-15otxrx6/overlay/lib/python3.10/site-packages/setuptools/build_meta.py", line 142, in run_setup
2050
      exec(compile(code, __file__, 'exec'), locals())
2051
    File "setup.py", line 50, in <module>
2052
      MAJOR, MINOR, MICRO = FULLVERSION.split('.')[:3]
2053
  ValueError: not enough values to unpack (expected 3, got 1)
```

File: .github/workflows/linux.yml [hunk 1]
```diff
@@ -56,32 +55,35 @@ jobs:
...
-      if: matrix.python-version == '3.9'
+    - name: Install Python 3.10
+      if: matrix.python-version == '3.10'
       run: |
         sudo add-apt-repository ppa:deadsnakes/ppa
         sudo apt-get update
-        sudo apt-get install -y --no-install-recommends python3.9-dev python3.9-distutils python3.9-venv
-        python3.9 -m pip install --upgrade pip setuptools
+        sudo apt install -y --no-install-recommends python3.10-dev python3.10-distutils python3.10-venv
+        curl -O https://bootstrap.pypa.io/get-pip.py && python3.10 get-pip.py && rm get-pip.py
```
Comment (rgommers) [inline .github/workflows/linux.yml:69]: So if this is necessary, it's a (temporary I guess) issue with the Python 3.10 that GitHub currently provides. Maybe add a comment that says that and points to this PR. We should be able to remove this later on.

**Comment by V0lantis** [comment]
Ok so for me, it ready to be merged

**Review by rgommers — APPROVED** [review]
All green, in it goes! Thanks @V0lantis

**Pull request merged**